dlt/common/schema/schema.py

```diff
@@ -177,6 +184,71 @@ def coerce_row(self, table_name: str, parent_table: str, row: StrAny) -> Tuple[D
 
         return new_row, updated_table_partial
 
+    def resolve_evolution_settings_for_table(self, parent_table: str, table_name: str, schema_evolution_settings_override: TSchemaEvolutionSettings) -> TSchemaEvolutionModes:
+
+        def resolve_single(settings: TSchemaEvolutionSettings) -> TSchemaEvolutionModes:
+            settings = settings or {}
+            if isinstance(settings, str):
+                return TSchemaEvolutionModes(table=settings, column=settings, column_variant=settings)
+            return settings
+
+        # find table settings
+        table_with_settings = parent_table or table_name
+
+        # modes
+        table_evolution_modes = resolve_single(self.tables.get(table_with_settings, {}).get("schema_evolution_settings", {}))
+        schema_evolution_modes = resolve_single(self._settings.get("schema_evolution_settings", {}))
+        overide_modes = resolve_single(schema_evolution_settings_override)
+
+        # resolve to correct settings dict
+        settings = cast(TSchemaEvolutionModes, {**DEFAULT_SCHEMA_EVOLUTION_MODES, **schema_evolution_modes, **table_evolution_modes, **overide_modes})
+
+        return settings
+
+
+    def check_schema_update(self, parent_table: str, table_name: str, row: DictStrAny, partial_table: TPartialTableSchema, schema_evolution_settings_override: TSchemaEvolutionSettings) -> Tuple[DictStrAny, TPartialTableSchema]:
+        """Checks if schema update mode allows for the requested changes, filter row or reject update, depending on the mode"""
+
+        assert partial_table
+
+        # for now we defined the schema as new if there are no data columns defined
+        has_columns = self.has_data_columns
+        if not has_columns:
+            return row, partial_table
+
+        evolution_modes = self.resolve_evolution_settings_for_table(parent_table, table_name, schema_evolution_settings_override)
+
+        # default settings allow all evolutions, skipp all else
+        if evolution_modes == DEFAULT_SCHEMA_EVOLUTION_MODES:
+            return row, partial_table
+
+        table_exists = table_name in self.tables and len(self.tables[table_name].get("columns", {}))
```

Comment:
use `get_table_columns` to make sure we allow only for complete columns. otherwise we can say that table is new

Reply:
also this code assumes that when we update schema, all partial tables for a single tables are merged first. I do not think it is the case.
you need to cache which table is new in the normalizer. otherwise you can add only single column to a new table (when new table is allowed but adding columns to existing table is not)

Reply:
Are you sure? I thought all the schema updates are merged into the schema after the normalizer finished running, at least that is what the code looks like to me. I also have the case of adding a new table with multiple columns when new columns are disabled in the test and it passes.

Reply:
you are right - but there are other places where schema is updated
1. schema is updated in extract: from hints but also from data (all hints may be lambdas - including columns)
2. each worker process in normalizer updates tables on the fly
```
                if partial_table:
                    # update schema and save the change
                    schema.update_schema(partial_table)
                    table_updates = schema_update.setdefault(table_name, [])
```
3. all partial tables are collected from all worker processes and "fresh" schema is updated again at the end of the normalization

how to test it
1. define a complete column in extract stage, it will get added to the table and passed to normalizer. in extract yield data which will create another column in normalizer (from data). that will IMO fail when you allow new tables but freeze columns creation
2. yield two rows. first row creates column1, second row creates column2 - that will also fail
3. make columns a lambda (dynamic hints) and yield two rows, for each rows lambda returns different column. that should already fail in extract phase